lib/charms/opensearch/v0/opensearch_locking.py

```diff
@@ -25,78 +25,240 @@
 logger = logging.getLogger(__name__)
 
 
-class OpenSearchOpsLock:
-    """This class covers the configuration changes depending on certain actions."""
-
-    LOCK_INDEX = ".ops_lock"
-    PEER_DATA_LOCK_FLAG = "ops_removing_unit"
+class _PeerRelationEndpoint(ops.Object):
+    _NAME = "node-lock-fallback"
 
-    def __init__(self, charm):
+    def __init__(self, charm: ops.CharmBase):
+        super().__init__(charm, self._NAME)
         self._charm = charm
-        self._opensearch = charm.opensearch
+        self.framework.observe(
+            self._charm.on[self._NAME].relation_changed, self._on_peer_relation_changed
+        )
+
+    @property
+    def acquired(self) -> bool:
+        """Attempt to acquire lock.
 
-    @retry(stop=stop_after_attempt(3), wait=wait_fixed(0.5), reraise=True)
-    def acquire(self):
-        """Method for Acquiring the "ops" lock."""
-        # no lock acquisition needed if only 1 unit remaining
-        if len(self._charm.model.get_relation(PeerRelationName).units) == 1:
+        Returns:
+            Whether lock was acquired
+        """
+        if not self._relation:
+            return False
+        self._relation.data[self._charm.unit]["lock-requested"] = json.dumps(True)
+        if self._charm.unit.is_leader():
+            logger.debug("[Node lock] Requested peer lock as leader unit")
+            # A separate relation-changed event won't get fired
+            self._on_peer_relation_changed()
+        acquired = self._unit_with_lock == self._charm.unit.name
+        if acquired:
+            logger.debug("[Node lock] Acquired via peer databag")
+        else:
+            logger.debug(
+                f"[Node lock] Not acquired. Unit with peer databag lock: {self._unit_with_lock}"
+            )
+        return acquired
+
+    def release(self):
+        """Release lock for this unit."""
+        if not self._relation:
             return
+        self._relation.data[self._charm.unit].pop("lock-requested", None)
+        if self._charm.unit.is_leader():
+            logger.debug("[Node lock] Released peer lock as leader unit")
+            # A separate relation-changed event won't get fired
+            self._on_peer_relation_changed()
+
+    def _unit_requested_lock(self, unit: ops.Unit):
+        """Whether unit requested lock."""
+        assert self._relation
+        value = self._relation.data[unit].get("lock-requested")
+        if not value:
+            return False
+        value = json.loads(value)
+        if not isinstance(value, bool):
+            raise ValueError
+        return value
+
+    @property
+    def _unit_with_lock(self):
+        if self._relation:
+            return self._relation.data[self._charm.app].get("unit-with-lock")
+
+    @_unit_with_lock.setter
+    def _unit_with_lock(self, value: str):
+        assert self._relation
```

Comment:
`assert` will throw an exception and I cannot see any case where you're handling that with a try/catch. I think this condition never *actually* happens anyways, but we should have a "if" and return None instead of an exception.

Reply:
@carlcsaposs-canonical we do not need the `assert` - as the subsequent `self._relation.data` will raise an exception anyway